src/Network/Response.php

```diff
@@ -2078,6 +2144,104 @@ public function file($path, array $options = [])
+     * - download: If `true` sets download header and forces file to
+     *   be downloaded rather than displayed inline.
+     *
+     * @param string $path Path to file. If the path is not an absolute path that resolves
+     *   to a file, `APP` will be prepended to the path.
+     * @param array $options Options See above.
+     * @return static
+     * @throws \Cake\Network\Exception\NotFoundException
+     */
+    public function withFile($path, array $options = [])
+    {
+        $file = $this->validateFile($path);
+        $options += [
+            'name' => null,
+            'download' => null
+        ];
+
+        $extension = strtolower($file->ext());
+        $mapped = $this->getMimeType($extension);
+        if ((!$extension || !$mapped) && $options['download'] === null) {
+            $options['download'] = true;
+        }
+
+        $new = clone $this;
+        if ($mapped) {
+            $new = $new->withType($extension);
+        }
+
+        $fileSize = $file->size();
+        if ($options['download']) {
+            $agent = env('HTTP_USER_AGENT');
+
+            if (preg_match('%Opera(/| )([0-9].[0-9]{1,2})%', $agent)) {
+                $contentType = 'application/octet-stream';
+            } elseif (preg_match('/MSIE ([0-9].[0-9]{1,2})/', $agent)) {
+                $contentType = 'application/force-download';
+            }
+
+            if (isset($contentType)) {
+                $new = $new->withType($contentType);
+            }
+            $name = $options['name'] ?: $file->name;
+            $new = $new->withDownload($name)
+                ->withHeader('Content-Transfer-Encoding', 'binary');
+        }
+
+        $new = $new->withHeader('Accept-Ranges', 'bytes');
+        $httpRange = env('HTTP_RANGE');
+        if (isset($httpRange)) {
+            $new->_fileRange($file, $httpRange);
+        } else {
+            $new = $new->withHeader('Content-Length', (string)$fileSize);
+        }
+        $new->_file = $file;
+        $new->stream = new Stream($file->path, 'rb');
+
+        return $new;
+    }
+
+    /**
+     * Validate a file path is a valid response body.
+     *
+     * @param string $path The path to the file.
+     * @throws \Cake\Network\Exception\NotFoundException
+     * @return \Cake\Filesystem\File
+     */
+    protected function validateFile($path)
+    {
+        if (strpos($path, '../') !== false || strpos($path, '..\\') !== false) {
+            throw new NotFoundException('The requested file contains `..` and will not be read.');
+        }
+        if (!is_file($path)) {
+            $path = APP . $path;
+        }
+
+        $file = new File($path);
+        if (!$file->exists() || !$file->readable()) {
+            if (Configure::read('debug')) {
+                throw new NotFoundException(sprintf('The requested file %s was not found or not readable', $path));
+            }
+            throw new NotFoundException(__d('cake', 'The requested file was not found'));
```

Comment:
The non-debug messages should be translated imho. Else it is required to hack to core to have a non-english software towards end users.